tiledb/sm/query/query_condition.cc

```diff
@@ -668,49 +695,130 @@ QueryCondition::apply_clause(
     case Datatype::STRING_UCS2:
     case Datatype::STRING_UCS4:
     default:
-      return {Status_QueryConditionError(
-                  "Cannot perform query comparison; Unsupported query "
-                  "conditional type on " +
-                  clause.field_name_),
-              nullopt};
+      throw std::runtime_error(
+          "QueryCondition::apply_ast_node: Cannot perform query comparison; "
+          "Unsupported query "
+          "conditional type on " +
+          node->get_field_name());
   }
 
-  return {Status::Ok(), nullopt};
+  return;
+}
+
+void QueryCondition::apply_tree(
```

Comment:
There is a lot of code duplication between `apply_tree` , `apply_tree_sparse` and `apply_tree_dense` . Correct me if I am wrong but they actually perform the exact same logic, but call recursively different flavors of apply in their bodies with slightly different arguments. 
This is an improvement we can do in a next PR, it will help reducing the surface of possible bugs.

Reply:
I think it would be good to combine these if possible -- we just recapitulated the previous structure to start with.

Reply:
I definitely agree with this! I may want to wait until this PR gets merged though to reconsider this. It may be in our best interest to create a story for this though!